modules/l4proxy/proxy.go

```diff
@@ -271,20 +270,35 @@ func (h *Handler) proxy(down *layer4.Connection, upConns []net.Conn) {
 		}(up)
 	}
 
-	// read from downstream until connection is closed;
-	// TODO: this pumps the reader, but writing into discard is a weird way to do it; could be avoided if we used io.Pipe - see _gitignore/oldtee.go.txt
-	io.Copy(ioutil.Discard, downTee)
+	downConnClosedCh := make(chan struct{}, 1)
```

Comment:
Is it necessary? 

because `wg.Wait()`  blocks until all upstream is closed. maybe just close downstream after `wg.Wait`?

And if downstream close before upstream, `io.Copy` in L263 will return, and `defer` in L261 will be executed.

Reply:
The basic idea here is to **gracefully close the downstream if the upstream is half-closed**, which is a common case in HTTP (see [9.6. Tear-down](https://httpwg.org/http-core/draft-ietf-httpbis-messaging-latest.html#persistent.tear-down)).

Suppose a scenario that consists of a downstream D (an HTTP client) and an upstream U (an HTTP server). If U performs a half-close (still can read from D), `io.Copy` in L263 will return, thus will finally make `wg.Wait` in L292 return. If we don't wait for the closure of D by `<-downConnClosedCh`, there's no chance for the last data from D to be read by U.

Half-closed connections are also common on the client side when involving tools such as netcat (#40) and FTP.

Reply:
Thanks for your explanation, Really appreciate it!